pkg/pillar/cmd/zedrouter/patchenvelopes.go

```diff
@@ -0,0 +1,254 @@
+		pes.log.Errorf("Failed to update external patches %v", err)
+	}
+}
+
+func (pes *PatchEnvelopes) deleteExternalPatches(vs types.VolumeStatus) {
+	pes.Lock()
+	defer pes.Unlock()
+
+	i := 0
+	for _, vol := range pes.completedVolumes {
+		if vol.VolumeID == vs.VolumeID {
+			break
+		}
+		i++
+	}
+
+	pes.completedVolumes[i] = pes.completedVolumes[len(pes.completedVolumes)-1]
+	pes.completedVolumes = pes.completedVolumes[:len(pes.completedVolumes)-1]
+}
+
+func (pes *PatchEnvelopes) updateEnvelopes(peInfo []types.PatchEnvelopeInfo) {
+	pes.Lock()
+	defer pes.Unlock()
+
+	pes.envelopes = peInfo
+	if err := pes.processVolumeStatusList(pes.completedVolumes); err != nil {
+		pes.log.Errorf("Failed to update external patches %v", err)
+	}
+}
+
+func (pes *PatchEnvelopes) processVolumeStatus(vs types.VolumeStatus) error {
+	for i := range pes.envelopes {
+		for j := range pes.envelopes[i].VolumeRefs {
+			volUUID, err := uuid.FromString(pes.envelopes[i].VolumeRefs[j].ImageID)
+			if err != nil {
+				return err
+			}
+			if volUUID == vs.VolumeID {
+				blob, err := blobFromVolRef(pes.envelopes[i].VolumeRefs[j], vs)
+				if err != nil {
+					return err
+				}
+
+				if idx := types.CompletedBinaryBlobIdxByName(pes.envelopes[i].BinaryBlobs, blob.FileName); idx != -1 {
+					pes.envelopes[i].BinaryBlobs[idx] = *blob
+				} else {
+					pes.envelopes[i].BinaryBlobs = append(pes.envelopes[i].BinaryBlobs, *blob)
+				}
+			}
+		}
+	}
+	return nil
+}
+
+func (pes *PatchEnvelopes) processVolumeStatusList(volumeStatuses []types.VolumeStatus) error {
+	for _, vs := range volumeStatuses {
+		if err := pes.processVolumeStatus(vs); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func blobFromVolRef(vr types.BinaryBlobVolumeRef, vs types.VolumeStatus) (*types.BinaryBlobCompleted, error) {
+	sha, err := fileutils.ComputeShaFile(vs.FileLocation)
+	if err != nil {
+		return nil, err
+	}
+	return &types.BinaryBlobCompleted{
+		FileName:         vr.FileName,
+		FileMetadata:     vr.FileMetadata,
+		ArtifactMetadata: vr.ArtifactMetadata,
+		FileSha:          fmt.Sprintf("%x", sha),
+		URL:              vs.FileLocation,
+	}, nil
+}
+
+// PatchEnvelopeInfo contains fields that we don't want to expose to app instance (like AllowedApps), so we use
+// peInfoToDisplay and patchEnvelopesJSONFOrAppInstance to marshal PatchEnvelopeInfoList in a format, which is
+// suitable for app instance.
+// We cannot use json:"-" structure tag to omit AllowedApps from json marshaling since we use PatchEnvelopeInfo between
+// zedagent and zedrouter to communicate new PatchEnvelopes from EdgeDevConfig. This communication is done via pubSub,
+// which uses json marshaling to communicate structures between processes. And using json:"-" will make AllowedApps "magically"
+// disappear on zedrouter
+type peInfoToDisplay struct {
+	PatchID     string
+	BinaryBlobs []types.BinaryBlobCompleted
+	VolumeRefs  []types.BinaryBlobVolumeRef
+}
+
+const metadataIP = "169.254.169.254"
```

Comment:
ELsewhere in zedrouter we use the same value I suspect. Makes sense to have a common const definition.

Reply:
moved it to zedrouter/metadata.go and used in other one occurrence